JoomGalleryfriends/JG4-dev
# IMGtools service

This PR fixes different bugs and makes a lot of code improvements on the IMGtools service.

## What is this service for
The IMGtools service will be used to perform image manipulations (resize, rotate, flip, orient, watermatk) and to handle image metadata (readMetadata, copyMetadata).

The service exists for two different image processors:
- GD (standard image processor delivered with PHP)
- ImageMagick (command line tool for image processing)
When loading the service you can choose which image processor you want to use.

## How to use this service
You can load the service anywhere in Joomla by doing
```
$component = Factory::getApplication()->bootComponent('com_joomgallery');  // boot JoomGallery component
$component->createIMGtools('gd');   // instantiate service with gd as image processor
$img = $component->getIMGtools();   // load the service into your scope
$img->read('image.jpg');   // read image from file
$img-><METHOD>   // whatever you want to do with the image
```

## How to test this service
Make sure "Debug System" is "Yes" in Joomla Global configuration.
Open the file `/administrator/com_joomgallery/src/View/Faulties/HtmlView.php` and insert the following code on line 62 after `throw new \Exception(implode("\n", $errors)); }`

```
// Test IMGtools
$this->component->createIMGtools('gd');
$img = $this->component->getIMGtools();
$img->set('keep_metadata', false);
$img->set('keep_anim', false);
$img->set('auto_orient', true);
$img->read('/images/<YOUR_IMAGE>.jpg');
$img->watermark('/images/joomla_black.png', 9, 2, 25, 50);
$img->resize(3, 500, 500);

// Output image
echo $img->stream(80);
$this->component->printDebug();

// Output image metadata
echo '<br /><br /><h3>Image metadata:</h3>';
dump($img->get('metadata'));
```
Navigate to Components->JoomGallery->Maintenance Manager to see the IMGtools service running...

## What to test
### Different Images
- Different images of type jpg, png and gif
- png and gif with and without transparency
- gif with and without animation
- jpg with metadata (especially exif orientation)

### Different processing methods
Test the images with different amount and different order of processing methods.
Available methods:
- resize($method, $width, $height, $cropposition=2, $unsharp=false)
- rotate($angle)
- flip($direction)
- orient()
- watermatk($wtm_file, $wtm_pos, $wtm_resize, $wtm_newSize, $opacity)

more infos: [IMGtools interface](https://github.com/Elfangor93/JG4-dev/blob/IMGtools/administrator/com_joomgallery/src/Service/IMGtools/IMGtoolsInterface.php)

### Different image processors
- gd (https://www.php.net/manual/en/book.image.php)
- im (https://imagemagick.org/script/convert.php)
